app/jobs/data_import_job.rb

```diff
@@ -5,53 +5,118 @@ class DataImportJob < ApplicationJob
   queue_as :low
 
   def perform(data_import)
-    contacts = []
-    data_import.update!(status: :processing)
-    csv = CSV.parse(data_import.import_file.download, headers: true)
-    csv.each { |row| contacts << build_contact(row.to_h.with_indifferent_access, data_import.account) }
-    result = Contact.import contacts, on_duplicate_key_update: :all, batch_size: 1000
-    data_import.update!(status: :completed, processed_records: csv.length - result.failed_instances.length, total_records: csv.length)
+    @data_import = data_import
+    process_import_file
+    send_failed_records_to_admin
   end
 
   private
 
-  def build_contact(params, account)
-    # TODO: rather than doing the find or initialize individually lets fetch objects in bulk and update them in memory
-    contact = init_contact(params, account)
+  def process_import_file
+    @data_import.update!(status: :processing)
+
+    contacts, rejected_contacts = parse_csv_and_build_contacts
+
+    import_contacts(contacts)
+    update_data_import_status(contacts.length, rejected_contacts.length)
+    save_failed_records_csv(rejected_contacts)
+  end
 
+  def parse_csv_and_build_contacts
+    contacts = []
+    rejected_contacts = []
+    csv = CSV.parse(@data_import.import_file.download, headers: true)
+
+    csv.each do |row|
+      current_contact = build_contact(row.to_h.with_indifferent_access, @data_import.account)
+      current_contact.valid? ? contacts << current_contact : rejected_contacts << current_contact
+    end
+
+    [contacts, rejected_contacts]
+  end
+
+  def import_contacts(contacts)
+    # <struct ActiveRecord::Import::Result failed_instances=[], num_inserts=1, ids=[444, 445], results=[]>
+    Contact.import(contacts, synchronize: contacts, on_duplicate_key_ignore: true, track_validation_failures: true, validate: true, batch_size: 1000)
+  end
+
+  def update_data_import_status(processed_records, rejected_records)
+    @data_import.update!(status: :completed, processed_records: processed_records, total_records: processed_records + rejected_records)
+  end
+
+  def build_contact(params, account)
+    contact = find_or_initialize_contact(params, account)
     contact.name = params[:name] if params[:name].present?
```

Comment:
Let's assign `company` , and `city` as well. Here in additional attributes. 
let's also look at whether we can add country as well.